### ISSUE TITLE
Support the use of Generic with __init_subclass__ arguments

### ISSUE THREAD

**Issue by Ovsyanka83** [microsoft/pyright#4171]
**Is your feature request related to a problem? Please describe.**
I have the following structure in my code:
```python
from typing import Generic, TypeVar


class BaseDatabaseModel:
    pass


Model = TypeVar("Model", bound=BaseDatabaseModel)


class SomeDatabaseModel(BaseDatabaseModel):
    pass


class BaseOperations(Generic[Model]):
    model: type[Model]

    def __init_subclass__(cls, model: type[Model]) -> None:
        cls.model = model


class BusinessLogic(BaseOperations, model=SomeDatabaseModel):
    pass


reveal_type(BusinessLogic().model)
```

Pyright is already able to see that `model` is a required parameter for any descendants of `BaseOperations` but sadly it doesn't support `Generic`. I.e. In this case `reveal_type` shows `Unknown` instead of `Type[SomeDatabaseModel]`. 

**Describe the solution you'd like**
`reveal_type` should show `Type[SomeDatabaseModel]`  similar to how it would do if we sent model through `__init__`. Pyright is great at guessing types and I know that you prefer for pyright to only support features that are supported by python's type system (and this one is) so it would be cool if you were able to implement it.

**Additional context**
Mypy gives `Type[Any]` for this `reveal_type` which is a bit more helpful but still not great.

**Comment by erictraut** [comment]
Pyright (and mypy) are doing the right thing here. The `BaseOperations` class is a generic class with one type parameter. The declaration of class `BusinessLogic` indicates that it derives from `BaseOperations`, but you have omitted the type argument for this class. The rules laid out by PEP 484 indicate that a type checker should treat missing type arguments as `Any`. (Pyright internally uses the variant type `Unknown` to indicate that it's an _implicit_ `Any`.) That means the class `BusinessLogic` derives from `BaseOperations[Any]`, and the type of instance variable `model` is therefore `type[Any]` (which is the same as `Any`).

If you declare `BusinessLogic` as follows, it will work correctly.

```python
class BusinessLogic(BaseOperations[SomeDatabaseModel], model=SomeDatabaseModel):
    pass
```
